optuna/importance/_base.py

```diff
@@ -0,0 +1,152 @@
+
+def get_distributions(study: Study, params: Optional[List[str]]) -> Dict[str, BaseDistribution]:
+    _check_evaluate_args(study, params)
+
+    if params is None:
+        return intersection_search_space(study, ordered_dict=True)
+
+    # New temporary required to pass mypy. Seems like a bug.
+    params_not_none = params
+    assert params_not_none is not None
+
+    # Compute the search space based on the subset of trials containing all parameters.
+    distributions = None
+    for trial in study.trials:
+        if trial.state != TrialState.COMPLETE:
+            continue
+
+        trial_distributions = trial.distributions
+        if not all(name in trial_distributions for name in params_not_none):
+            continue
+
+        if distributions is None:
+            distributions = dict(
+                filter(
+                    lambda name_and_distribution: name_and_distribution[0] in params_not_none,
+                    trial_distributions.items(),
+                )
+            )
+            continue
+
+        if any(
+            trial_distributions[name] != distribution
+            for name, distribution in distributions.items()
+        ):
+            raise ValueError(
+                "Parameters importances cannot be assessed with dynamic search spaces if "
+                "parameters are specified. Specified parameters: {}.".format(params)
+            )
+
+    assert distributions is not None  # Requires to pass mypy.
+    distributions = OrderedDict(
+        sorted(distributions.items(), key=lambda name_and_distribution: name_and_distribution[0])
+    )
+    return distributions
+
+
+def get_study_data(
+    study: Study, distributions: Dict[str, BaseDistribution]
+) -> Tuple[np.ndarray, np.ndarray]:
+    trials = []
+    for trial in study.trials:
+        if trial.state != TrialState.COMPLETE:
+            continue
+        if any(name not in trial.params for name in distributions.keys()):
+            continue
+        trials.append(trial)
+
+    n_trials = len(trials)
+    n_params = len(distributions)
+
+    params = np.empty((n_trials, n_params), dtype=np.float64)
+    values = np.empty((n_trials,), dtype=np.float64)
+
+    for i, trial in enumerate(trials):
+        trial_params = trial.params
+        for j, (name, distribution) in enumerate(distributions.items()):
+            param = trial_params[name]
+            if isinstance(distribution, CategoricalDistribution):
+                param = distribution.to_internal_repr(param)
+            params[i, j] = param
+        values[i] = trial.value
+
+    return params, values
+
+
+def _check_evaluate_args(study: Study, params: Optional[List[str]]) -> None:
+    completed_trials = list(filter(lambda t: t.state == TrialState.COMPLETE, study.trials))
+    if len(completed_trials) == 0:
+        raise ValueError("Cannot evaluate parameter importances without completed trials.")
+
+    if params is not None:
+        if not isinstance(params, (list, tuple)):
```

Comment:
We don't have to, but I think it's quite common among Python users to expect a simple enumeration (container of strings) to also accept a tuple. Do you think it's better to be strict?

Reply:
Thanks. That makes sense. It seems okay to keep this code as it is.

Reply:
**FYI** I found that `Trial.suggest_categorical` uses `Sequence[...]` to allow lists and tuples (and other sequence types).

https://github.com/optuna/optuna/blob/b72426ffd138d1d4770077f6e45186c15a37b900/optuna/trial.py#L47-L48

Reply:
I see, thanks for the note. `Sequence` contains 'str' types as well so I'm not sure it's appropriate here nor in `Trial.suggest_categorical` 🤔 .

Reply:
Hmm, `suggest_categorical` uses `Sequence[CategoricalChoiceType]`, a specialized type of `Sequence`,  as the type of `choices` argument. I'm not sure but it seems differ from `str`(maybe this is another specialized type of `Sequence`?).

For instance, I applied `mypy` to the following code:
```python
from typing import Sequence

def foo(seq: Sequence[int]) -> None:
    pass

foo("hello")
```

Then, `mypy` emitted the following error:
```
foo.py:8: error: Argument 1 to "foo" has incompatible type "str"; expected "Sequence[int]"
Found 1 error in 1 file (checked 1 source file)
```

So, it seems `mypy` recognizes as `Sequence[int]` and `str` are different types.

Reply:
Note that I wrote the above comment just to clarity the behavior of type hint and `mypy`. I think that it's okay to keep using `List[str]` as the type of `param` argument because list type may be more familiar with normal users and sufficient for typical use cases.

Reply:
Ah, and if you don't pass any arguments to the type hint of `Sequence`, `mypy` don't complain when passing `str`s. That makes sense. Thank you for the clarification. However, I agree that a simple `List` is more user friendly.